### PR TITLE
(psa) make workloads compatible with psa:restricted profile

### LINKED ISSUE #2644: [FLAKY] cleanup csvs with bad owner operator groups
### Description

Unfortunately, I marked this as a flake without capturing the stack trace ._.
Once I get it I'll update the issue. Or if this test fails for you, please paste in the stack trace and log link if possible.

### PR DESCRIPTION
**Pull request by anik120**
**Motivation for the change:**

With the introduction of [Pod Security Admission](https://kubernetes.io/docs/concepts/security/pod-security-admission/#pod-security-admission-labels-for-namespaces), the recommended best practice is to enforce the `Restricted` policy of admission (see [1] for more details).

[1] https://kubernetes.io/docs/concepts/security/pod-security-standards/#restricted

**Description of the change:**

This PR
    *) Lables the olm namespace as `enforce:restricted`
    *) Labels the operators namespace as `enforce:baseline` (to allow existing CSV deployments
    without securityContext set to deploy in the namespace, which won't be possible with
    `enforce:resticted`)
    *) updates the securityContext of olm workload pods(olm-operator, catalog-operator,
    and CatalogSource registry pods) to adhere to the `Restricted` policy.
    *) updates the bundle unpacking job to create a pod that adheres to the `Restricted` policy,
    so that bundles can be unpacked in the `Restricted` namespace

Will also close #2644 as a byproduct of the need to fix test failures 

**Architectural changes:**

**Testing remarks:**

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage
- [ ] Sufficient end-to-end test coverage
- [ ] Bug fixes are accompanied by regression test(s)
- [ ] e2e tests and flake fixes are accompanied evidence of flake testing, e.g. executing the test 100(0) times
- [ ] tech debt/todo is accompanied by issue link(s) in comments in the surrounding code
- [ ] Tests are comprehensible, e.g. Ginkgo DSL is being used appropriately
- [ ] Docs updated or added to `/doc`
- [ ] Commit messages sensible and descriptive
- [ ] Tests marked as `[FLAKE]` are truly flaky and have an issue
- [ ] Code is properly formatted

### CODE REVIEW

File: deploy/chart/templates/_packageserver.deployment-spec.yaml
Comment (perdasilva) [inline deploy/chart/templates/_packageserver.deployment-spec.yaml:21]: nice!

File: deploy/chart/templates/0000_50_olm_00-namespace.yaml
Comment (perdasilva) [inline deploy/chart/templates/0000_50_olm_00-namespace.yaml:6]: we should also add
`pod-security.kubernetes.io/restricted-version: latest`

File: deploy/chart/templates/0000_50_olm_00-namespace.yaml
Comment (perdasilva) [inline deploy/chart/templates/0000_50_olm_00-namespace.yaml:14]: same as above

**Comment by perdasilva** [comment]
In general it lgtm. Seems to hit all the right changes. Nicely done.
In case I missed it, it's probably worthwhile adding input validation to the runAsUser. I think at the moment 0 is allowed (as input), but won't end up stamping out a runAsUser (which is fine). But, without the validation it could be a bit confusing for the user.

I still suggest not giving the userId as a degree of freedom to the admin, at least not if we don't have a specific ask for it. The reasons being is that it could lead to backwards compatibility pain later on in case we need to fix the userId for some reason or another. If there's no clear customer need, we're also increasing the complexity of the code for no good reason. Even if we do needed it, I wonder if there's a nice library or pattern in go for storing this kind of configuration so it doesn't have to get passed all the way down the stack.

**Comment by perdasilva** [comment]
Another couple of things that occured to me while I slept is that by default we _need_ to have the runAsUser in the securityContext of the jobs and catalog source pods in the majority of cases. The reason being that we cannot guarantee that the catalog source image or the bundle image have users defined in their containers. But, we kept the default userId value to -1 (i.e. no runAsUser). This means that the new flag is always necessary (except in OCP where SCC will inject the value). Even if we still want to keep the user id flag (rather than use a toggle), we should probably have a different default so that it doesn't have to be used/defined in the majority case.

**Comment by anik120** [comment]
>Another couple of things that occured to me while I slept is that by default we need to have the runAsUser in the securityContext of the jobs and catalog source pods in the majority of cases. The reason being that we cannot guarantee that the catalog source image or the bundle image have users defined in their containers. But, we kept the default userId value to -1 (i.e. no runAsUser). This means that the new flag is always necessary (except in OCP where SCC will inject the value). Even if we still want to keep the user id flag (rather than use a toggle), we should probably have a different default so that it doesn't have to be used/defined in the majority case.

I see your point, I've changed the default value to 1001. 

> In case I missed it, it's probably worthwhile adding input validation to the runAsUser. I think at the moment 0 is allowed (as input), but won't end up stamping out a runAsUser (which is fine). But, without the validation it could be a bit confusing for the user.

With that ☝🏽 change in the context, the validation  `if runAsUser >= 0{ set UserID}` allows for any positive numerical value, including 0 (root), which is fine since if someone sets `runAsNonRoot` but sets `runAsUser` to 0, PSA controller will help us out by denying admission. But also, if for whatever reason, we do need to change back to `runAsRoot`, we can easily set UID to 0 without any code change. 

>I still suggest not giving the userId as a degree of freedom to the admin, at least not if we don't have a specific ask for it. The reasons being is that it could lead to backwards compatibility pain later on in case we need to fix the userId for some reason or another.

We do have other args that we pass to the containers, and the same backward compatibility argument can be extended to them too no? In fact, if we hardcode the value, and have to change the value, we'll have to actually backport code changes. Whereas if we don't hardcode the values, and we need to change the UID for previous releases, all we need is release manifest changes for previous releases 🎉 

> If there's no clear customer need, we're also increasing the complexity of the code for no good reason

On the contrary, if we don't leave this degree of freedom, besides the problem mentioned above, we'll possibly have to keep changing the hardcoded value with evolving PSA enablement effort org wide at present too. 

Hardcoding "run workload as UID" sounds like brings the exact problems you're trying to avoid in the first place @perdasilva , and the solution is to not harcode it.

**Comment by anik120** [comment]
Oh and 
> Even if we do needed it, I wonder if there's a nice library or pattern in go for storing this kind of configuration so it doesn't have to get passed all the way down the stack.

Unfortunately, since the workloads are being created way down in the stack, the configuration needs to be passed down all the way to the bottom (almost). I am not a Go expert yet (😉) either, but this is the pattern I've always observed being used every where. And I believe it is one of the short comings of Go too, but don't quote me on that (says he after documenting it down in writing), although I'm sure an actual Go expert can articulate the shortcoming better than me.

**Comment by perdasilva** [comment]
It's more a question of control. If we need to change it, we can, and it's not impact for the user. If we decide we need to pin to a user or another because of the container or other possible future changes, we can easily control that. Up and downstream. I'd just check that we aren't pulling from olm:latest for the controller. If that's not the case then the point is moot. But, just as a general guideline (and that's all it is), don't add more functionality than you need, because you'll end up having to support it.

**Comment by perdasilva** [comment]
Oh, just came to me. Something we'd do a lot at Amazon for our services was to use env vars instead. Those can be easily picked up the code without having to traverse the whole stack. It also avoids all the unit testing issues of using a singleton.

**Review by camilamacedo86 — CHANGES_REQUESTED** [review]
Hi @anik120 that is great but we can **not** use **RunAsUser** 
We need to remove ALL RunAsUser fields added.


#### Option A)  

Define the USER ID via the Dockerfile image. Therefore, we do NOT need to use the spec RunAsUser in the SecurityContext spec  at all and we do not face the issue "`container has runAsNonRoot and image will run as root` "

Example:
Docker Image
USER 1001

Then, SecurityContext:

```ymal
    spec:
      securityContext:
        runAsNonRoot: true
        # Please ensure that you can use SeccompProfile and do not use
        # if your project must work on old Kubernetes
        # versions < 1.19 or on vendors versions which
        # do NOT support this field by default (i.e. Openshift < 4.11 )
        seccompProfile:
          type: RuntimeDefault
      ...
      containers:
      - name: controller-manager
        securityContext:
          allowPrivilegeEscalation: false
          capabilities:
            drop:
            - ALL

```

Result: the containers will be qualified to restricted-v2 in OCP and will run as restricted on k8s

####  Option B: 

- We will need to ensure that the namespace is enforced with the Baseline
- We do not use the RunAsNonRoot must be empty or false

File: cmd/catalog/start.go
Comment (camilamacedo86) [inline cmd/catalog/start.go:70]: Why are we adding this flag? 
What is the image used which does not define the USER ID?

File: deploy/chart/templates/0000_50_olm_00-namespace.yaml
Comment (camilamacedo86) [inline deploy/chart/templates/0000_50_olm_00-namespace.yaml:8]:
```suggestion
```
If we enforce: restricted then we do not need to warn it.
The containers that cannot run as restricted will not be admitted so has no point to warn

File: deploy/chart/templates/0000_50_olm_00-namespace.yaml
Comment (camilamacedo86) [inline deploy/chart/templates/0000_50_olm_00-namespace.yaml:19]:
```suggestion
    pod-security.kubernetes.io/warn: restricted
```

If the enforce is baseline then we can warn when any container on this namespace is not restricted OR not set up the warning at all. But the warning makes only sense if be less permissive than the enforcement.

File: deploy/chart/templates/0000_50_olm_07-olm-operator.deployment.yaml
Comment (camilamacedo86) [inline deploy/chart/templates/0000_50_olm_07-olm-operator.deployment.yaml:26]:
```suggestion
```
We cannot use RunAsUser at all,
That is ONLY required because the Dockerfile image is not defining the user when should be.
So, what is the image used here?

File: deploy/chart/templates/0000_50_olm_08-catalog-operator.deployment.yaml
Comment (camilamacedo86) [inline deploy/chart/templates/0000_50_olm_08-catalog-operator.deployment.yaml:26]:
```suggestion
```
Same. We cannot use RunAsUser at all,
That is ONLY required because the Dockerfile image is not defining the user when should be.
So, what is the image used here? 

Another option, in this case, is to ensure enforcement as a baseline or privileged and not set RunAsNonRoot either.

File: deploy/chart/templates/0000_50_olm_08-catalog-operator.deployment.yaml
Comment (camilamacedo86) [inline deploy/chart/templates/0000_50_olm_08-catalog-operator.deployment.yaml:90]:
```suggestion
```
The idea of passing the USER ID will not work.
We have no way to know a fixed value number in OCP.

File: deploy/chart/templates/0000_50_olm_08-catalog-operator.deployment.yaml
Comment (perdasilva) [reply deploy/chart/templates/0000_50_olm_08-catalog-operator.deployment.yaml:26]: Nah we should keep security as nerfed as possible. I don't think the controllers need anything above restricted.

File: deploy/chart/templates/0000_50_olm_08-catalog-operator.deployment.yaml
Comment (perdasilva) [reply deploy/chart/templates/0000_50_olm_08-catalog-operator.deployment.yaml:90]: The idea here for the downstream is to omit the flag and let SCC inject the user id in the range it wants.

File: deploy/chart/templates/0000_50_olm_07-olm-operator.deployment.yaml
Comment (perdasilva) [reply deploy/chart/templates/0000_50_olm_07-olm-operator.deployment.yaml:26]: That's a fair point actually. Since we are baking the USER into the olm container we don't need runAsUser for the controller deployments. Only for the catalog source and unpack jobs, since we have less control there.

File: deploy/chart/templates/0000_50_olm_00-namespace.yaml
Comment (perdasilva) [reply deploy/chart/templates/0000_50_olm_00-namespace.yaml:19]: Oh, we should enforce restricted on the olm namespace! Can remove the warn (and audit) label(s)

**Comment by perdasilva** [comment]
Regarding validation it's better to fail fast, rather than let PSA catch it. Might cause confusion. Idk that I'd change the runAsNonRoot behavior. More stuff to test.

File: cmd/catalog/start.go
Comment (camilamacedo86) [inline cmd/catalog/start.go:70]:
```suggestion
```


The idea of passing the USER ID will not work.
We do have not a way to know a fixed value number valid for OCP. 
Also, that would mean ensuring that the ServiceAccount used would have permissions to the anyuid SCC.

Therefore, we have two options:
a) Ensure that the  USERID is define in the dockerfile image
OR 
b) Do not set ANY userID and do NOT set RunAsNonRoot 
Then, for downstream we need to ensure that the SA has access to the required SCC ( i.e. previleged )
Also, ensure that the namespace will be labeled


The valid range is based on the namespace config. 
Also, we cannot use RunAsUser. 

From OCP:

> 

If your S2I image does not include a USER declaration with a numeric user, your builds will fail by default. In order to allow images that use either named users or the root (0) user to build in OpenShift Container Platform, you can add the project’s builder service account (system:serviceaccount:<your-project>:builder) to the privileged security context constraint (SCC). Alternatively, you can allow all images to run as any user.
--

File: deploy/chart/templates/_packageserver.deployment-spec.yaml
Comment (camilamacedo86) [inline deploy/chart/templates/_packageserver.deployment-spec.yaml:23]:
```suggestion
```
Same

File: deploy/chart/values.yaml
Comment (camilamacedo86) [inline deploy/chart/values.yaml:57]:
```suggestion
```
Same

Comment (camilamacedo86) [reply deploy/chart/values.yaml:57]: The image `quay.io/operator-framework/olm:master` must define the USER ID instead

File: pkg/controller/bundle/bundle_unpacker.go
Comment (camilamacedo86) [inline pkg/controller/bundle/bundle_unpacker.go:108]: will we run this version in OCP < 4.11 ? If so we cannot use it.

File: pkg/controller/bundle/bundle_unpacker.go
Comment (camilamacedo86) [inline pkg/controller/bundle/bundle_unpacker.go:223]:
```suggestion
```
Same

File: pkg/controller/bundle/bundle_unpacker.go
Comment (camilamacedo86) [inline pkg/controller/bundle/bundle_unpacker.go:259]:
```suggestion
	UnpackBundle(lookup *operatorsv1alpha1.BundleLookup, timeout time.Duration) (result *BundleUnpackResult, err error)
```

File: pkg/controller/bundle/bundle_unpacker.go
Comment (camilamacedo86) [inline pkg/controller/bundle/bundle_unpacker.go:413]:
```suggestion
func (c *ConfigMapUnpacker) UnpackBundle(lookup *operatorsv1alpha1.BundleLookup, timeout time.Duration) (result *BundleUnpackResult, err error) {
```

File: pkg/controller/bundle/bundle_unpacker.go
Comment (camilamacedo86) [inline pkg/controller/bundle/bundle_unpacker.go:475]:
```suggestion
	job, err = c.ensureJob(cmRef, result.Path, secrets, timeout)
```

File: pkg/controller/bundle/bundle_unpacker.go
Comment (camilamacedo86) [inline pkg/controller/bundle/bundle_unpacker.go:615]:
```suggestion
	fresh := c.job(cmRef, bundlePath, secrets, timeout)
```

File: pkg/controller/bundle/bundle_unpacker_test.go
Comment (camilamacedo86) [inline pkg/controller/bundle/bundle_unpacker_test.go:227]:
```suggestion
```

File: pkg/controller/bundle/bundle_unpacker_test.go
Comment (camilamacedo86) [inline pkg/controller/bundle/bundle_unpacker_test.go:36]:
```suggestion
```

**Comment by anik120** [comment]
> Hi @anik120 that is great but we can **not** use **RunAsUser** We need to remove ALL RunAsUser fields added.
> 
> #### Option A)  
> Define the USER ID via the Dockerfile image. Therefore, we do NOT need to use the spec RunAsUser in the SecurityContext spec at all and we do not face the issue "`container has runAsNonRoot and image will run as root` "
> 
> Example: Docker Image USER 1001
> 
> Then, SecurityContext:
> 
> ```
>     spec:
>       securityContext:
>         runAsNonRoot: true
> ```

@camilamacedo86 keep in mind that this is the upstream PR. Conversations about setting `runAsUser` to comply with OCP SCC are not valid for this PR. That'll be discussed while downstreaming this PR.

**Comment by perdasilva** [comment]
> On the contrary, if we don't leave this degree of freedom, besides the problem mentioned above, we'll possibly have to keep changing the hardcoded value with evolving PSA enablement effort org wide at present too.

I'm more worried about upstream users here. Changing a hard-coded value is trivial. Deprecating behavior is hard. Removing the degree of freedom here gives us control to do what we want without affecting users or having to take support promises into account. I don't know what kinds of decisions admins will make around user id and how removing ability would affect them. But I'll leave it to you to decide. It's just a call out. Configuration certainly has its advantages. I just cannot predict how this space will change, which is why I tend to fall on the side of keeping control.

File: deploy/chart/values.yaml
Comment (camilamacedo86) [reply deploy/chart/values.yaml:57]: @anik120 We need to define the user in : https://github.com/operator-framework/operator-lifecycle-manager/blob/master/Dockerfile#L36-L38

File: pkg/controller/bundle/bundle_unpacker.go
Comment (anik120) [reply pkg/controller/bundle/bundle_unpacker.go:108]: We will not, this is for upstream: latest. Please reserve all conversations about OCP for the downstream PR.

File: cmd/catalog/start.go
Comment (anik120) [reply cmd/catalog/start.go:70]: Again, this conversation is valid for downstream, so we'll discuss this in the downstream PR.

File: deploy/chart/templates/0000_50_olm_08-catalog-operator.deployment.yaml
Comment (anik120) [reply deploy/chart/templates/0000_50_olm_08-catalog-operator.deployment.yaml:90]: +1 to what Per said.

**Review by camilamacedo86 — CHANGES_REQUESTED** [review]
We need to change the Dockerfile images instead of pass the UserId as flag
/hold

File: deploy/chart/templates/0000_50_olm_08-catalog-operator.deployment.yaml
Comment (camilamacedo86) [reply deploy/chart/templates/0000_50_olm_08-catalog-operator.deployment.yaml:90]: > The idea here for the downstream is to omit the flag and let SCC inject the user id in the range it wants\

I understood the idea and that will not work.
We have no way to know the value to inform. What value will you pass when it will run in OCP? That is not a field value or a fixed range at all. 

Also, the good practices for Docker are to have the userID set in the image ( which sort out the issue ), remove all these complexities, and make it eaiser for us to keep the project maintained.

File: deploy/chart/templates/0000_50_olm_07-olm-operator.deployment.yaml
Comment (camilamacedo86) [inline deploy/chart/templates/0000_50_olm_07-olm-operator.deployment.yaml:26]:
```suggestion
```

File: deploy/chart/templates/0000_50_olm_08-catalog-operator.deployment.yaml
Comment (camilamacedo86) [reply deploy/chart/templates/0000_50_olm_08-catalog-operator.deployment.yaml:90]: @perdasilva @anik120 could you describe how this injection will work?

File: deploy/chart/templates/0000_50_olm_00-namespace.yaml
Comment (camilamacedo86) [reply deploy/chart/templates/0000_50_olm_00-namespace.yaml:8]: @anik120 ^

File: deploy/chart/templates/0000_50_olm_00-namespace.yaml
Comment (camilamacedo86) [reply deploy/chart/templates/0000_50_olm_00-namespace.yaml:19]: @anik120

File: deploy/chart/templates/0000_50_olm_07-olm-operator.deployment.yaml
Comment (camilamacedo86) [reply deploy/chart/templates/0000_50_olm_07-olm-operator.deployment.yaml:26]: @anik120

File: deploy/chart/templates/0000_50_olm_07-olm-operator.deployment.yaml
Comment (camilamacedo86) [inline deploy/chart/templates/0000_50_olm_07-olm-operator.deployment.yaml:25]: @anik120 @perdasilva,
 
In "downstream" if we omit this value when the image (Dockerfile) does not set the USER ID that will fail with the error `container has runAsNonRoot and image has non-numeric user`.

Please, test this scenario with an image that requires the RunAsUserID example: 'memcached:1.4.36-alpine'
Example : https://gist.github.com/camilamacedo86/c7c06e51272f8c348655123d0f18eceb

File: deploy/chart/templates/0000_50_olm_00-namespace.yaml
Comment (anik120) [reply deploy/chart/templates/0000_50_olm_00-namespace.yaml]: For the `operators` namespace, I've updated it to be the following: 

```
pod-security.kubernetes.io/enforce: baseline
pod-security.kubernetes.io/enforce-version: latest
pod-security.kubernetes.io/warn: privileged
pod-security.kubernetes.io/warn-version: latest
```

From what I've read, I think this means "allow baseline pods, but also allow privileged pod with a warning". Reason for keeping this configuration is that baseline should allow most of the CSVs out there to be installed in that namespace. If there are any operators that require privileged access, we need to allow those, but throw a warning. PLMK if the above configuration doesn't achieve what I think it does.

**Comment by anik120** [comment]
@perdasilva @camilamacedo86 updated the PR with the following updates: 

1. The olm and catalog operator deployment does not explicitly set `runAsUser`, instead the images are built with user 1001. 
2. The registry pods and the bundle unpacking job *still absolutely needs* their `runAsUser` set explicitly, since they're not being built from any image (check code of how they're created for more details) 
3. The flag for the `catalog-operator` container arg is not a boolean (that sets the `runAsUser` as mentioned in 2 to 1001 if set to `true`)
>I'm more worried about upstream users here. Changing a hard-coded value is trivial. Deprecating behavior is hard. Removing the degree of freedom here gives us control to do what we want without affecting users or having to take support promises into account

That makes sense to me. It'll behoove us to take this conversation to the wg meeting and let the community know that we are exposing this flag, and based on the feedback we receive, if we see that there's appetite for needing to set a custom user id for those workloads, we can introduce a new flag and get rid of the boolean one.

File: deploy/chart/templates/0000_50_olm_00-namespace.yaml
Comment (camilamacedo86) [inline deploy/chart/templates/0000_50_olm_00-namespace.yaml:15]: @anik120 @perdasilva any pod installed on this namespace will NOT have ONLY restricted-v2 SCC. 
Therefore, if you do not inform the RunAsUser and you set RunAsNonRoot true the SCC will not inject the RunAsUser. 

It will fail in Openshift when the image does not define the UserID

Comment (camilamacedo86) [reply deploy/chart/templates/0000_50_olm_00-namespace.yaml:15]: /hold

File: deploy/chart/templates/0000_50_olm_00-namespace.yaml
Comment (camilamacedo86) [inline deploy/chart/templates/0000_50_olm_00-namespace.yaml:17]: @anik120 @perdasilva 

This namespace has `pod-security.kubernetes.io/enforce: baseline` which means `privileged ` pods will not run. So, make no sense we add the ` pod-security.kubernetes.io/warn: privileged`

File: deploy/chart/templates/0000_50_olm_00-namespace.yaml
Comment (camilamacedo86) [inline deploy/chart/templates/0000_50_olm_00-namespace.yaml:6]: In Openshift the label syncer can change that. 
Therefore, that means the solution proposed here will fail when:

- The Dockerfile image does not define the USER ID
AND
- The Pod be created with RunAsNonRoot
AND
- The label syncer changes the enforcement of this namespace  

@perdasilva @anik120

**Review by camilamacedo86 — COMMENTED** [review]
/lgtm

**Review by camilamacedo86 — COMMENTED** [review]
/hold

File: deploy/chart/templates/0000_50_olm_00-namespace.yaml
Comment (camilamacedo86) [inline deploy/chart/templates/0000_50_olm_00-namespace.yaml:16]: @anik120 is just a detail, if we add here the latest that means it will get the latest criteria available from the policy. therefore, in the future, it might no longer work with the changes in this code/release.

Comment (camilamacedo86) [reply deploy/chart/templates/0000_50_olm_00-namespace.yaml:16]: we create an issue to address this scenario: https://github.com/operator-framework/operator-lifecycle-manager/issues/2827 and it is blocking the release and will get done as follow up so all fine. 

/hold cancel

**Pull request merged**
